intellij/src/saros/intellij/negotiation/ModuleConfigurationInitializer.java

```diff
@@ -81,43 +95,163 @@ public void enqueueModuleConfigurationChange(
 
-    Map<String, String> configurationOptions = queuedModuleOptions.remove(module);
+    ModuleConfiguration moduleConfiguration = queuedModuleOptions.remove(module);
 
-    if (configurationOptions == null) {
+    if (moduleConfiguration == null) {
       log.debug("Skipped module root model adjustment as not options were found - " + module);
 
       return;
     }
 
-    ModuleRootModificationUtil.updateModel(
-        module,
-        modifiableRootModel -> applyModuleConfiguration(modifiableRootModel, configurationOptions));
+    if (!moduleConfiguration.isExistingModule()) {
+      ModuleRootModificationUtil.updateModel(
+          module,
+          modifiableRootModel ->
+              applyModuleConfiguration(modifiableRootModel, moduleConfiguration));
+
+    } else {
+      ContentEntry[] contentEntries = ModuleRootManager.getInstance(module).getContentEntries();
+
+      if (contentEntries.length != 1) {
+        log.error("Encountered shared module with multiple content roots - " + module);
+
+        return;
+      }
+
+      ContentEntry contentEntry = contentEntries[0];
+
+      if (configurationDiffers(contentEntry, moduleConfiguration.getRootPaths())) {
+
+        Runnable changeModuleConfiguration =
+            () ->
+                ModuleRootModificationUtil.updateModel(
+                    module,
+                    modifiableRootModel -> {
+                      resetModuleConfiguration(modifiableRootModel);
+                      applyModuleConfiguration(modifiableRootModel, moduleConfiguration);
+                    });
+
+        SafeDialogUtils.showYesNoDialog(
+            module.getProject(),
+            Messages.ModuleConfigurationInitializer_override_module_config_message,
+            Messages.ModuleConfigurationInitializer_override_module_config_title,
+            changeModuleConfiguration);
+      }
+    }
   }
 
   /**
-   * Applies the additional module options shared as part of the module project negotiation data to
-   * the given modifiable module root model.
+   * Returns whether the content entry source root configuration matches the passed configuration.
    *
-   * <p>The available options are defined through keys in {@link ModuleConfigurationProvider}.
+   * @param contentEntry the content entry to check
+   * @param rootPaths the root paths received from the host
+   * @return <code>true</code> if the root paths of the passed content entry match the paths
+   *     received from the host, <code>false</code> otherwise
+   */
+  private boolean configurationDiffers(
+      @NotNull ContentEntry contentEntry,
+      @NotNull Map<JpsModuleSourceRootType<? extends JpsElement>, String[]> rootPaths) {
+
+    for (Entry<JpsModuleSourceRootType<? extends JpsElement>, String[]> entry :
+        rootPaths.entrySet()) {
+
+      JpsModuleSourceRootType<? extends JpsElement> type = entry.getKey();
+      String[] paths = entry.getValue();
+
+      if (paths == null) {
```

Comment:
Use empty set instead. Might differ even if no received. (case: local has config, host does not)

Reply:
Done.